### PR TITLE
xds: Add RLS Cluster Specifier Plugin

### PR DESCRIPTION
**Pull request by zasweq**
This PR adds the RLS Cluster Specifier Plugin. This is branched off #4987. The RLS Cluster Specifier Plugin will not be functional until the RLS LB Policy itself gets completed, as right now the call into the RLS LB Policies ParseConfig() for validation purposes is commented out.

RELEASE NOTES: None

### CODE REVIEW

File: xds/internal/clusterspecifier/rls/rls.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+type rls struct{}
+
+func (rls) TypeURLs() []string {
+	return []string{"type.googleapis.com/grpc.lookup.v1.RouteLookupClusterSpecifier"}
+}
+
+// LBConfigJSON is the RLS LB Policies configuration in JSON format.
+// RouteLookupConfig will be a raw JSON string from the passed in proto
+// configuration, and the other fields will be hardcoded.
+type LBConfigJSON struct {
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:46]: This type shouldn't need to be exported.

File: xds/internal/clusterspecifier/rls/rls.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	RouteLookupConfig                interface{}              `json:"routeLookupConfig"`
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:47]: `[]byte`?

File: xds/internal/clusterspecifier/rls/rls.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	any, ok := cfg.(*anypb.Any)
+	if !ok {
+		return nil, fmt.Errorf("rls_csp: error parsing config %v: unknown type %T", cfg, cfg)
+	}
+	rlcs := new(grpc_lookup_v1.RouteLookupClusterSpecifier)
+
+	if err := ptypes.UnmarshalAny(any, rlcs); err != nil {
+		return nil, fmt.Errorf("rls_csp: error parsing config %v: %v", cfg, err)
+	}
+	// Do we use protojson or regular marshal function - see if Marshal produces correct results?
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:65]: Why wouldn't we use `protojson`?  Even if `json.Marshal` happens to work given what's currently there, `protojson` is what you're supposed to use to marshal proto into JSON and vice-versa.

File: xds/internal/clusterspecifier/rls/rls.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	// if err != nil {
+	//	return nil, fmt.Errorf("rls_csp: error marshaling load balancing config %v: %v", lbCfgJSON, err)
+	// }
+
+	// _, err := rls.ParseConfig(rawJSON) (will this function need to change its logic at all?)
+	// if err != nil {
+	// return nil, fmt.Errorf("error: %v", err)
+	// }
+
+	return []map[string]interface{}{{"rls_experimental": lbCfgJSON}}, nil
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:94]:
```go
return clusterspecifier.BalancerConfig{{"rls_experimental": lbCfgJSON}}, nil
```

File: xds/internal/clusterspecifier/rls/rls.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+	// rawJSON, err := json.Marshal(lbCfgJSON)
+	// if err != nil {
+	//	return nil, fmt.Errorf("rls_csp: error marshaling load balancing config %v: %v", lbCfgJSON, err)
+	// }
+
+	// _, err := rls.ParseConfig(rawJSON) (will this function need to change its logic at all?)
+	// if err != nil {
+	// return nil, fmt.Errorf("error: %v", err)
+	// }
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:92]: `rls.ParseConfig` exists, actually.  Can we make it accessible and use it?

Also, it would be used via:

```go
rlsBB := balancer.Get("experimental_rls")
if rlsBB == nil {
	return nil, fmt.Errorf("RLS LB policy not registered")
}
rlsBB.(balancer.ConfigParser).ParseConfig(.....)
```

File: xds/internal/clusterspecifier/rls/rls.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	rlcJSON, err := json.Marshal(rlcs.GetRouteLookupConfig())
+	if err != nil {
+		return nil, fmt.Errorf("rls_csp: error marshaling route lookup config: %v: %v", rlcs.GetRouteLookupConfig(), err)
+	}
+
+	lbCfgJSON := &LBConfigJSON{
+		RouteLookupConfig: rlcJSON, // "JSON form of RouteLookupClusterSpecifier.config" - RLS in xDS Design Doc
+		ChildPolicy: []map[string]interface{}{
+			{
+				"cds_experimental": struct{}{}, // Is this right? {} in design doc
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:76]: Does `nil` work?

File: xds/internal/xdsclient/xdsresource/unmarshal_rds_test.go [hunk 7]
```diff
@@ -727,6 +759,15 @@ func (errorClusterSpecifierPlugin) ParseClusterSpecifierConfig(proto.Message) (c
...
 }
 
+var rlsClusterSpecifierConfig = testutils.MarshalAny(&grpc_lookup_v1.RouteLookupClusterSpecifier{
+	// Once Easwar merges the RLS LB implementation and the rls csp actually
+	// calls into it for validation, this config will have to be scaled up to
+	// pass the validations present in the RLS LB implementation.
+	RouteLookupConfig: &grpc_lookup_v1.RouteLookupConfig{
+		LookupService: "rls-specifier",
+	},
+})
```
Comment (dfawley) [inline xds/internal/xdsclient/xdsresource/unmarshal_rds_test.go:769]: I'm not sure this is the right way to test this.

The xdsclient tests shouldn't rely on any specific plugins, and the tests for any specific plugins should not be done in the xdsclient tests.  Instead, the rls plugin's tests should be done in `xds/internal/clusterspecifier/rls`.

Comment (zasweq) [reply xds/internal/xdsclient/xdsresource/unmarshal_rds_test.go:769]: Hmmmmmm ok noted. Switched to package, and tested input/output of ParseClusterSpecifierConfig().

File: xds/internal/clusterspecifier/rls/rls.go (hunk 5 above)
Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls.go:92]: Done. However, what is specified as the output of nil in your other comment doesn't seem to work. This is because it causes a nil map in rls ParseConfig (then parsing Child Policy config). I will try and fix it by making it just "{}"?

**Review by zasweq — COMMENTED** [review]
Got to all your comments, but commit is still a WIP.

File: xds/internal/clusterspecifier/rls/rls.go (hunk 6 above)
Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls.go:76]: Tried a literal bytestring []byte(`{}`) and also didn't work. This needs to unmarshal into map[string]json.RawMessage (https://github.com/grpc/grpc-go/blob/master/balancer/rls/internal/config.go#L273). Any suggestions? Is {} a valid config for cds_experimental?

Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls.go:76]: I marshaled {} into a json.RawMessage to pass RLS Config Parsing.

File: xds/internal/xdsclient/xdsresource/unmarshal_rds_test.go (hunk 7 above)
Comment (zasweq) [reply xds/internal/xdsclient/xdsresource/unmarshal_rds_test.go]: Actually, manually testing and comparing the output (containing long marshaled json.RawMessage) understandably led to some flakiness. I pushed the fix in a separate commit, that way you can see what I had previously.

File: xds/internal/clusterspecifier/rls/rls.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	rlsBB := balancer.Get("rls")
+	if rlsBB == nil {
+		return nil, fmt.Errorf("RLS LB policy not registered")
+	}
+	_, err = rlsBB.(balancer.ConfigParser).ParseConfig(rawJSON)
+	if err != nil {
+		return nil, fmt.Errorf("rls_csp: validation error from rls lb policy parsing %v", err)
+	}
+
+	return clusterspecifier.BalancerConfig{{"rls_experimental": lbCfgJSON}}, nil
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:99]: I was going to suggest using an exported constant in the RLS package, but I don't think we can do that for visibility reasons without moving stuff around invasively.

Will the tests break when there is no "rls_experimental" LB policy anymore (it will eventually be renamed to "rls")?  If so, then I think this is fine.

File: xds/internal/clusterspecifier/rls/rls.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package rls implements the RLS cluster specifier plugin.
+package rls
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:20]: Should this be importing the RLS package?  Seems like it should.

File: xds/internal/clusterspecifier/rls/rls.go [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/golang/protobuf/ptypes"
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/internal/proto/grpc_lookup_v1"
+	"google.golang.org/grpc/xds/internal/clusterspecifier"
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/anypb"
+)
+
+func init() {
+	clusterspecifier.Register(rls{})
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:36]: Should this be conditional on `GRPC_EXPERIMENTAL_XDS_RLS_LB`?

File: xds/internal/clusterspecifier/rls/rls.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+	rlcs := new(grpc_lookup_v1.RouteLookupClusterSpecifier)
+
+	if err := ptypes.UnmarshalAny(any, rlcs); err != nil {
+		return nil, fmt.Errorf("rls_csp: error parsing config %v: %v", cfg, err)
+	}
+	rlcJSON, err := protojson.Marshal(rlcs.GetRouteLookupConfig())
+	if err != nil {
+		return nil, fmt.Errorf("rls_csp: error marshaling route lookup config: %v: %v", rlcs.GetRouteLookupConfig(), err)
+	}
+	emptyJSON, err := json.Marshal(struct{}{})
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:71]: Isn't this just `json.RawMessage("{}")`?  If so, let's use that (inline below) instead of something that can error.

File: xds/internal/clusterspecifier/rls/rls.go [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+		},
+		ChildPolicyConfigTargetFieldName: "cluster",
+	}
+
+	rawJSON, err := json.Marshal(lbCfgJSON)
+	if err != nil {
+		return nil, fmt.Errorf("rls_csp: error marshaling load balancing config %v: %v", lbCfgJSON, err)
+	}
+
+	rlsBB := balancer.Get("rls")
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls.go:90]: Do this as a global?

Also...shouldn't this match `"rls_experimental"` below?  Can you use a single const for both of these?

File: xds/internal/clusterspecifier/rls/rls_test.go [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+			name: "valid-rls-cluster-specifier",
+			rlcs: rlsClusterSpecifierConfigWithoutTransformations,
+		},
+	}
+	for _, test := range tests {
+		cs := clusterspecifier.Get("type.googleapis.com/grpc.lookup.v1.RouteLookupClusterSpecifier")
+		if cs == nil {
+			t.Fatal("Error getting cluster specifier")
+		}
+		_, err := cs.ParseClusterSpecifierConfig(test.rlcs)
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls_test.go:65]: Validate the result?  (I'm assuming it would be marshaling the result to JSON and compare to a string.)

File: xds/internal/clusterspecifier/rls/rls.go (hunk 9 above)
Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls.go:20]: I did, but it failed vet. Let me try again.

File: xds/internal/clusterspecifier/rls/rls.go (hunk 10 above)
Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls.go:36]: Should I do this in another PR and also protect the xdsclient changes?

File: xds/internal/clusterspecifier/rls/rls.go (hunk 12 above)
Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls.go:90]: This is the name it's registered by in balancer/rls/internal. Below corresponds to what is defined in design doc.

File: xds/internal/clusterspecifier/rls/rls.go (hunk 11 above)
Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls.go:71]: Switched.

File: xds/internal/clusterspecifier/rls/rls.go (hunk 8 above)
Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls.go:99]: This is named rls in master. I'm going to switch what gets spit out of this to rls, that way once it gets to balancer group, it actually pulls it off.

File: xds/internal/clusterspecifier/rls/rls.go (hunk 9 above)
Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls.go:20]: But I agree, that way the import dependency graph through the xdsclient gets this balancer eventually registered.

**Review by zasweq — COMMENTED** [review]
Whoops, these were sitting in pending.

File: xds/internal/clusterspecifier/rls/rls_test.go (hunk 13 above)
Comment (dfawley) [reply xds/internal/clusterspecifier/rls/rls_test.go]: I was expecting to see something like this here now:

```go
lbCfgJSON, err := json.Marshal(lbCfg)
// handle err
var got interface{}
err := json.Unmarshal(lbCfgJSON, &got)
// handle err
wantJSON := `
{
	// JSON that we wanted
}
`
var want interface{}
err := json.Unmarshal(wantJSON, &want)
// handle err
if diff := cmp.Diff(got, want); diff != "" {
	// error
}
```

File: xds/internal/clusterspecifier/rls/rls.go (hunk 8 above)
Comment (dfawley) [reply xds/internal/clusterspecifier/rls/rls.go]: As discussed, please rename to `rls_experimental` throughout.

File: xds/internal/clusterspecifier/rls/rls_test.go (hunk 13 above)
Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls_test.go]: Done. Although I kept the want as a struct, and then Marshaled and then Unmarshaled. I think it's cleaner.

File: xds/internal/clusterspecifier/rls/rls.go (hunk 8 above)
Comment (zasweq) [reply xds/internal/clusterspecifier/rls/rls.go]: Switched.

File: balancer/rls/internal/builder.go [hunk 14]
```diff
@@ -24,7 +24,7 @@ import (
 	"google.golang.org/grpc/internal/grpcsync"
 )
 
-const rlsBalancerName = "rls"
+const rlsBalancerName = "rls_experimental"
```
Comment (dfawley) [inline balancer/rls/internal/builder.go:27]: Looks like this was changed on master; hopefully rebasing will go smoothly.

File: xds/internal/clusterspecifier/rls/rls_test.go [hunk 15]
```diff
@@ -75,19 +76,31 @@ func (s) TestParseClusterSpecifierConfig(t *testing.T) {
...
-		if !reflect.DeepEqual(lbCfgJSON, wantCfgJSON) {
-			t.Fatalf("ParseClusterSpecifierConfig(%+v) = %v, want %v", test.rlcs, lbCfgJSON, wantCfgJSON)
+		var want interface{}
+		err = json.Unmarshal(wantCfgJSON, want)
+		if err != nil {
+			t.Fatalf("json.Unmarshal(%+v) returned err %v", lbCfgJSON, err)
+		}
+		if !cmp.Equal(want, got, cmpopts.EquateEmpty()) {
+			t.Fatalf("ParseClusterSpecifierConfig(%+v) returned expected, diff (-want +got):\\n%s", test.rlcs, cmp.Diff(want, got, cmpopts.EquateEmpty()))
 		}
```
Comment (dfawley) [inline xds/internal/clusterspecifier/rls/rls_test.go:104]: Use `if diff := cmp.Diff(want, got, cmpopts.EquateEmpty()); diff != "" {` and then there's no need to call `Diff` in the error message.

**Review by dfawley — APPROVED** [review]
